### PR TITLE
Exu style

### PR DESCRIPTION
**Pull request by bkorpan**
Style refactor of exu package.

### CODE REVIEW

File: src/main/scala/common/parameters.scala [hunk 1]
```diff
@@ -249,14 +249,14 @@ trait HasBoomCoreParameters extends freechips.rocketchip.tile.HasCoreParameters
    val STQ_ADDR_SZ       = log2Ceil(NUM_STQ_ENTRIES)
    val LSU_ADDR_SZ       = LDQ_ADDR_SZ max STQ_ADDR_SZ
    val BR_TAG_SZ         = log2Ceil(MAX_BR_COUNT)
-   val NUM_BROB_ENTRIES  = NUM_ROB_ROWS //TODO explore smaller BROBs
-   val BROB_ADDR_SZ      = log2Ceil(NUM_BROB_ENTRIES)
+   val NUM_BROB_ENTRIES  = numRobRows //TODO explore smaller BROBs
```
Comment (abejgonzalez) [inline src/main/scala/common/parameters.scala:252]: These are unused. Delete them?

File: src/main/scala/common/micro-op.scala [hunk 2]
```diff
@@ -83,7 +83,7 @@ class MicroOp(implicit p: Parameters) extends BoomBundle()(p)
    val imm_packed       = UInt(LONGEST_IMM_SZ.W) // densely pack the imm in decode...
```
Comment (abejgonzalez) [inline src/main/scala/common/micro-op.scala:83]: Just to confirm, the `common` refactor will happen later?

File: src/main/scala/exu/core.scala [hunk 3]
```diff
@@ -50,21 +50,21 @@ import boom.util.{GetNewUopAndBrMask, Sext, WrapInc}
...
-         val ptw_tlb = new freechips.rocketchip.rocket.TLBPTWIO()
-         val trace = Output(Vec(coreParams.retireWidth,
-            new freechips.rocketchip.rocket.TracedInstruction))
-         val release = Flipped(Valid(new boom.lsu.ReleaseInfo))
-         val fcsr_rm = UInt(freechips.rocketchip.tile.FPConstants.RM_SZ.W)
-   }
+  implicit val p: Parameters
+  val io = new freechips.rocketchip.tile.CoreBundle()(p)
+    with freechips.rocketchip.tile.HasExternallyDrivenTileConstants
+  {
```
Comment (abejgonzalez) [inline src/main/scala/exu/core.scala:56]: Do we want this on the previous line at the end?

File: src/main/scala/exu/core.scala [hunk 4]
```diff
@@ -50,21 +50,21 @@ import boom.util.{GetNewUopAndBrMask, Sext, WrapInc}
...
-         val ptw = Flipped(new freechips.rocketchip.rocket.DatapathPTWIO())
-         val rocc = Flipped(new freechips.rocketchip.tile.RoCCCoreIO())
-         val ptw_tlb = new freechips.rocketchip.rocket.TLBPTWIO()
-         val trace = Output(Vec(coreParams.retireWidth,
-            new freechips.rocketchip.rocket.TracedInstruction))
-         val release = Flipped(Valid(new boom.lsu.ReleaseInfo))
-         val fcsr_rm = UInt(freechips.rocketchip.tile.FPConstants.RM_SZ.W)
-   }
+  implicit val p: Parameters
+  val io = new freechips.rocketchip.tile.CoreBundle()(p)
```
Comment (abejgonzalez) [inline src/main/scala/exu/core.scala]: I believe you can get rid of the `()(p)` in all the places that this is used. It makes it look clearer and neater.

File: src/main/scala/exu/core.scala [hunk 5]
```diff
@@ -73,1489 +73,1386 @@ trait HasBoomCoreIO extends freechips.rocketchip.tile.HasTileParameters
...
+  val numFpWakeupPorts        = if (usingFPU) fp_pipeline.io.wakeups.length else 0
+
+  val decode_units     = for (w <- 0 until decodeWidth) yield { val d = Module(new DecodeUnit); d }
+  val dec_brmask_logic = Module(new BranchMaskGenerationLogic(coreWidth))
+  val rename_stage     = Module(new RenameStage(coreWidth, numIntRenameWakeupPorts, numFpWakeupPorts))
+  val issue_units      = new boom.exu.IssueUnits(numIntIssueWakeupPorts)
+  val dispatcher       = Module(new BasicDispatcher)
+
+  val iregfile         = if (enableCustomRf) {
+                           Module(new RegisterFileSeqCustomArray(numIntPhysRegs,
```
Comment (abejgonzalez) [inline src/main/scala/exu/core.scala]: IMO. The `numIntPhysRegs` should be on the next line.

File: src/main/scala/exu/core.scala [hunk 6]
```diff
@@ -73,1489 +73,1386 @@ trait HasBoomCoreIO extends freechips.rocketchip.tile.HasTileParameters
...
+                             numIrfReadPorts,
+                             numIrfWritePorts + 1, // + 1 for ll writebacks
+                             xLen,
+                             Seq(true) ++ exe_units.bypassable_write_port_mask)) // 0th is bypassable ll_wb
+                         }
+  val ll_wbarb         = Module(new Arbiter(new ExeUnitResp(xLen), 1 +
+                                                                   (if (usingFPU) 1 else 0) +
+                                                                   (if (usingRoCC) 1 else 0)))
+  val iregister_read   = Module(new RegisterRead(
+                                issue_units.map(_.issueWidth).sum,
```
Comment (abejgonzalez) [inline src/main/scala/exu/core.scala]: The `iregister_read` and `rob` modules should be...

```
Module....
  <- Indent
```

File: src/main/scala/exu/core.scala [hunk 7]
```diff
@@ -73,1489 +73,1386 @@ trait HasBoomCoreIO extends freechips.rocketchip.tile.HasTileParameters
...
+  val debug_irt_reg = RegInit(0.U(xLen.W))
+  debug_tsc_reg := debug_tsc_reg + Mux(O3PIPEVIEW_PRINTF.B, O3_CYCLE_TIME.U, 1.U)
+  debug_irt_reg := debug_irt_reg + PopCount(rob.io.commit.valids.asUInt)
+  dontTouch(debug_tsc_reg)
+  dontTouch(debug_irt_reg)
+
+  //****************************************
+  // Print-out information about the machine
+
+  val iss_str =
```
Comment (abejgonzalez) [inline src/main/scala/exu/core.scala]: The following variables should be camelCase: iss_str, exe_units_str, fp_pipeline_str, rob_str

File: src/main/scala/exu/decode.scala [hunk 8]
```diff
@@ -53,75 +53,53 @@ abstract trait DecodeConstants
...
-      val sigs =
-         Seq(legal, fp_val, fp_single, uopc, iqtype, fu_code, dst_type, rs1_type,
-             rs2_type, frs3_en, imm_sel, is_load, is_store, is_amo,
-             is_fence, is_fencei, mem_cmd, mem_typ, wakeup_delay, bypassable,
-             br_or_jmp, is_jal, allocate_brtag, is_sys_pc2epc, inst_unique, flush_on_commit, csr_cmd)
+  val legal           = Bool()
+  val fp_val          = Bool()
+  val fp_single       = Bool()
+  val uopc            = UInt(UOPC_SZ.W)
+  val iqtype          = UInt(IQT_SZ.W)
```
Comment (abejgonzalez) [inline src/main/scala/exu/decode.scala]: I think we should match this with the other `type` variables aka `iq_type`

File: src/main/scala/exu/decode.scala [hunk 9]
```diff
@@ -587,21 +585,21 @@ class DecodeUnit(implicit p: Parameters) extends BoomModule()(p)
...
-      val is_jalr = Output(Bool())
-      val is_ret  = Output(Bool())
-      val is_call = Output(Bool())
-      val target = Output(UInt(vaddrBitsExtended.W))
-      val cfi_type = Output(UInt(CfiType.SZ.W))
-   })
-
-   val bpd_csignals =
-      freechips.rocketchip.rocket.DecodeLogic(io.inst,
+  val io = IO(new Bundle
```
Comment (abejgonzalez) [inline src/main/scala/exu/decode.scala]: Similar to the other bundle comment, I think this { should be on the same line.

Comment (abejgonzalez) [reply src/main/scala/exu/decode.scala]: Maybe add to the style guide to do one or the other.

File: src/main/scala/exu/execution-units/execution-unit.scala [hunk 10]
```diff
@@ -35,574 +35,560 @@ import boom.util.{ImmGen, IsKilledByBranch, BranchKillableQueue}
...
-   val num_bypass_ports: Int,
-   val data_width: Int
-   )(implicit p: Parameters) extends BoomBundle()(p)
+  val writesIrf        : Boolean,
+  val writesLlIrf      : Boolean,
+  val writesFrf        : Boolean,
+  val writesLlFrf      : Boolean,
+  val hasRocc          : Boolean,
+  val hasBrUnit        : Boolean,
+  val has_fcsr         : Boolean,
```
Comment (abejgonzalez) [inline src/main/scala/exu/execution-units/execution-unit.scala]: Is there a reason why `has_fcsr and numbypassports` are not converted?

File: src/main/scala/exu/execution-units/execution-unit.scala [hunk 11]
```diff
@@ -35,574 +35,560 @@ import boom.util.{ImmGen, IsKilledByBranch, BranchKillableQueue}
...
+    io.iresp.valid     := iresp_fu_units.map(_.io.resp.valid).reduce(_|_)
+    io.iresp.bits.uop  := PriorityMux(iresp_fu_units.map(f =>
+      (f.io.resp.valid, f.io.resp.bits.uop.asUInt))).asTypeOf(new MicroOp())
+    io.iresp.bits.data := PriorityMux(iresp_fu_units.map(f =>
+      (f.io.resp.valid, f.io.resp.bits.data.asUInt))).asUInt
+
+    // pulled out for critical path reasons
+    // TODO: Does this make sense as part of the iresp bundle?
+    if (hasAlu) {
+       io.iresp.bits.uop.csr_addr := ImmGen(alu.io.resp.bits.uop.imm_packed, IS_I).asUInt
```
Comment (abejgonzalez) [inline src/main/scala/exu/execution-units/execution-unit.scala]: Indentation

File: src/main/scala/exu/execution-units/execution-units.scala [hunk 12]
```diff
@@ -26,187 +26,161 @@ import boom.common._
...
+    val fp_width = issueParams.find(_.iqType == IQT_FP.litValue).get.issueWidth
+    for (w <- 0 until fp_width) {
+      val fpu_exe_unit = Module(new FPUExeUnit(hasFpu = true,
+                                             hasFdiv = usingFDivSqrt && (w==0),
+                                             hasFpiu = (w==0)))
+      exe_units += fpu_exe_unit
+    }
+  }
+
+  val exe_units_str = new StringBuilder
```
Comment (abejgonzalez) [inline src/main/scala/exu/execution-units/execution-units.scala]: CamelCase

File: src/main/scala/exu/fp-pipeline.scala [hunk 13]
```diff
@@ -28,236 +28,227 @@ import boom.common._
...
+  {
+    val brinfo           = Input(new BrResolutionInfo())
+    val flush_pipeline   = Input(Bool())
+    val fcsr_rm          = Input(UInt(width=freechips.rocketchip.tile.FPConstants.RM_SZ.W))
+
+    val dis_uops         = Vec(coreWidth, Flipped(Decoupled(new MicroOp)))
+
+    // +1 for recoding.
+    val ll_wport         = Flipped(Decoupled(new ExeUnitResp(fLen+1)))// from memory unit
+    val fromint          = Flipped(Decoupled(new ExeUnitResp(fLen+1)))// from integer RF
```
Comment (abejgonzalez) [inline src/main/scala/exu/fp-pipeline.scala]: do we want to add a `_` to these names

File: src/main/scala/exu/register-read/regfile-custom.scala [hunk 14]
```diff
@@ -23,64 +23,56 @@ import freechips.rocketchip.config.Parameters
...
+  } else {
+    Module(new RegisterFileArray(numRegisters, numReadPorts, numWritePorts, registerWidth))
   }
   regfile.io.clock := clock
 
   // decode addrs into OH's
-  val waddr_OH = Wire(Vec(num_write_ports, UInt(num_registers.W)))
-  val raddr_OH = Reg(Vec( num_read_ports, UInt(num_registers.W)))
+  val waddr_OH = Wire(Vec(numWritePorts, UInt(numRegisters.W)))
+  val raddr_OH = Reg(Vec( numReadPorts, UInt(numRegisters.W)))
```
Comment (abejgonzalez) [inline src/main/scala/exu/register-read/regfile-custom.scala]: Spacing is a bit off.

File: src/main/scala/exu/rename/rename-busytable.scala [hunk 15]
```diff
@@ -61,165 +61,154 @@ class BusyTableIo(
...
+    busy_table.io.prs(0,w) := io.map_table(w).prs1
+    busy_table.io.prs(1,w) := io.map_table(w).prs2
+
+    io.values(w).prs1_busy := io.ren_uops(w).lrs1_rtype === rtype.U &&
+                              (busy_table.io.prs_busy(0,w) || prs1_was_bypassed(w))
+    io.values(w).prs2_busy := io.ren_uops(w).lrs2_rtype === rtype.U &&
+                              (busy_table.io.prs_busy(1,w) || prs2_was_bypassed(w))
+
+    if (rtype == RT_FLT.litValue) {
+       busy_table.io.prs(2,w) := io.map_table(w).prs3
```
Comment (abejgonzalez) [inline src/main/scala/exu/rename/rename-busytable.scala]: Indentation

File: src/main/scala/exu/rename/rename-busytable.scala [hunk 16]
```diff
@@ -61,165 +61,154 @@ class BusyTableIo(
...
+
+  // figure out if we need to bypass a newly allocated physical register from a previous instruction in this cycle.
+  val prs1_was_bypassed = WireInit(VecInit(Seq.fill(plWidth) {false.B}))
+  val prs2_was_bypassed = WireInit(VecInit(Seq.fill(plWidth) {false.B}))
+  val prs3_was_bypassed = WireInit(VecInit(Seq.fill(plWidth) {false.B}))
+  for {
+    w <- 0 until plWidth
+    xx <- w-1 to 0 by -1
+  }{
+    when (io.ren_uops(w).lrs1_rtype === rtype.U && io.ren_will_fire(xx) && io.ren_uops(xx).ldst_val &&
```
Comment (abejgonzalez) [inline src/main/scala/exu/rename/rename-busytable.scala:165]: IMO put the `}{` on the previous line.

File: src/main/scala/exu/rename/rename-busytable.scala [hunk 17]
```diff
@@ -61,165 +61,154 @@ class BusyTableIo(
...
+          io.ren_uops(xx).dst_rtype === rtype.U && (io.ren_uops(w).lrs1 === io.ren_uops(xx).ldst))
+      { prs1_was_bypassed(w) := true.B }
+    when (io.ren_uops(w).lrs2_rtype === rtype.U && io.ren_will_fire(xx) && io.ren_uops(xx).ldst_val &&
+          io.ren_uops(xx).dst_rtype === rtype.U && (io.ren_uops(w).lrs2 === io.ren_uops(xx).ldst))
+      { prs2_was_bypassed(w) := true.B }
+
+    if (rtype == RT_FLT.litValue) {
+       when (io.ren_uops(w).frs3_en && io.ren_will_fire(xx) && io.ren_uops(xx).ldst_val &&
```
Comment (abejgonzalez) [inline src/main/scala/exu/rename/rename-busytable.scala:173]: Indentation

File: src/main/scala/exu/rob.scala [hunk 18]
```diff
@@ -20,7 +20,7 @@
 //
 // NOTES:
 //    - Currently we do not compress out bubbles in the ROB.
-//    - commit_width is tied directly to the dispatch_width.
+//    - commit_coreWidth is tied directly to the dispatchWidth.
```
Comment (abejgonzalez) [inline src/main/scala/exu/rob.scala]: ?

**Comment by abejgonzalez** [comment]
My comments in general are...

Bundles should be either of the two options:
```
val io = IO(new Bundle {
  ...
}

val io = IO(new Bundle
{
}
```
I started doing the prior one in my `bpu` changes but we can reverse to the 2nd option if we want.

Additionally, I would remove all `(p)` references. This is already handled through the implicit parameter Scala support so we can make all the instantiations like this (see BPU for a second look at this):
```
Something()(p) ->> Something
Something(.......)(p) ->> Something(....)
```

File: src/main/scala/common/micro-op.scala (hunk 2 above)
Comment (bkorpan) [reply src/main/scala/common/micro-op.scala:83]: yes

File: src/main/scala/exu/execution-units/execution-unit.scala (hunk 10 above)
Comment (bkorpan) [reply src/main/scala/exu/execution-units/execution-unit.scala]: Rebase conflicts probably.

File: src/main/scala/exu/fp-pipeline.scala (hunk 13 above)
Comment (bkorpan) [reply src/main/scala/exu/fp-pipeline.scala]: yes

File: src/main/scala/exu/rob.scala (hunk 18 above)
Comment (bkorpan) [reply src/main/scala/exu/rob.scala]: Find & replace artifact.

File: src/main/scala/exu/rename/rename-busytable.scala (hunk 16 above)
Comment (bkorpan) [reply src/main/scala/exu/rename/rename-busytable.scala:165]: I disagree, I like a clear distinction between the multiline loop setup and body.

File: src/main/scala/exu/core.scala (hunk 3 above)
Comment (bkorpan) [reply src/main/scala/exu/core.scala:56]: In this particular case I'm not a fan of having the "with" at the same indentation as the following declarations, but its not a big deal.

**Review by abejgonzalez — APPROVED** [review]

File: src/main/scala/exu/core.scala (hunk 3 above)
Comment (abejgonzalez) [reply src/main/scala/exu/core.scala:56]: Likewise but I don't want to create exceptions

File: src/main/scala/exu/rename/rename-busytable.scala (hunk 16 above)
Comment (abejgonzalez) [reply src/main/scala/exu/rename/rename-busytable.scala:165]: Ok. Sounds good then.

**Pull request merged**